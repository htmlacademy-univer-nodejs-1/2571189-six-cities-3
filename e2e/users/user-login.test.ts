import assert from 'node:assert';
import { Headers, fetch } from 'undici';
import { describe, test } from 'vitest';
import LoggedUserRdo from '../../src/modules/user/rdo/logged-user.rdo.js';

process.env['E2E_ENDPOINT'] = 'http://localhost:4000';
assert(process.env['E2E_ENDPOINT'] !== undefined);
assert(process.env['E2E_ENDPOINT'].startsWith('http') === true, 'E2E_ENDPOINT должен начинаться с протокола');

const url = new URL(process.env['E2E_ENDPOINT']);

describe('POST /users/login', async () => {
  test('Success user login', async (tc) => {
    const user = {
      email: 'test@test.com',
      password: '12345678910',
    };

    await fetch(new URL('/users/register', url), {
      method: 'POST',
      headers: new Headers({
        'content-type': 'application/json',
      }),
      body: JSON.stringify({
        type: 'simple',
        email: 'test@test.com',
        username: 'name',
        password: '12345678910',
      })
    });

    const response = await fetch(new URL('/users/login', url), {
      method: 'POST',
      headers: new Headers({
        'content-type': 'application/json',
      }),
      body: JSON.stringify(user)
    });

    tc.expect(response.ok).toBeTruthy();
    tc.expect(response.status).toStrictEqual(200);
    tc.expect(response.headers.get('content-type')).toMatch(/application\/json/);
    const res = await response.json() as LoggedUserRdo;
    tc.expect(res.email).toStrictEqual(user.email);
    tc.expect(res.token.length).toBeGreaterThan(0);
  });
  test('Invalid password or email', async (tc) => {
    const user = {
      email: 'test@test.com',
      password: 'invalid',
    };

    const response = await fetch(new URL('/users/login', url), {
      method: 'POST',
      headers: new Headers({
        'content-type': 'application/json',
      }),
      body: JSON.stringify(user)
    });

    tc.expect(response.ok).toBeFalsy();
    tc.expect(response.status).toStrictEqual(401);
    tc.expect(response.statusText).toStrictEqual('Unauthorized');
    tc.expect(response.headers.get('content-type')).toMatch(/application\/json/);
    tc.expect(await response.json()).toMatchSnapshot();
  });
});
