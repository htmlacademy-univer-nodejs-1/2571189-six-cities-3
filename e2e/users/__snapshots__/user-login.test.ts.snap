// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`POST /users/login > Invalid password or email 1`] = `
{
  "error": "Unauthorized",
}
`;
